### ISSUE TITLE
Corrupt PDF from G4-encoded TIFF

### ISSUE THREAD

**Issue by carygravel** [PhilterPaper/Perl-PDF-Builder#167]
[Here](https://github.com/PhilterPaper/Perl-PDF-Builder/files/6707571/g4.zip) is another TIFF that PDF::Builder doesn't embed correctly.

I'll work on this now. I'm guessing the problem is that the TIFF is compressed in strips, and PDF has no concept of strips. There was the same problem with LZW.

**Comment by PhilterPaper** [comment]
I can confirm that the PDF from this TIFF fails to display in all my readers: Adobe Acrobat, Firefox, Chrome, XpdfReader. Thanks for taking a look at this -- I appreciate your work on TIFF. Don't forget to check if the PlanarConfig might be different in this case.

**Comment by carygravel** [comment]
The tests runs fine everywhere except on Windows:

https://github.com/carygravel/Perl-PDF-Builder/runs/3923782560?check_suite_focus=true

I'm guessing that this means that the newly wrapped TIFFReverseBits() is segfaulting for some reason, although the windows tests for Graphics::TIFF itself pass fine:

https://github.com/carygravel/graphics-tiff/actions/runs/1328737417

Would you mind installing Graphics::TIFF v17, run tiff.t from PDF::Builder, and give me the error message?

**Comment by carygravel** [comment]
> With v17, all 19 t/tiff.t tests return "ok". Could you check that you're up-to-date on PDF::Builder? Do I need to pull #174 first?

Yes. I wouldn't merge it just yet, just pull to a local branch, please and test again.

**Comment by PhilterPaper** [comment]
Reopening, as discussion taking place here for PRs such as #175.

**Comment by PhilterPaper** [comment]
test.yml reports Full Windows failed on Run choco install...
```
Ghostscript.app v9.55.0 [Approved]
ghostscript.app package files install completed. Performing other installation steps.
Installing 64 bit version
Installing Ghostscript.app...
Ghostscript.app has been installed.
  ghostscript.app may be able to be automatically uninstalled.
ghostscript not installed. An error occurred during installation:
 Access to the path 'C:\ProgramData\chocolatey\lib\Ghostscript.app\tools\gs9550w64.exe' is denied.
ghostscript package files install failed with exit code 1. Performing other installation steps.
The install of ghostscript was NOT successful.
ghostscript not installed. An error occurred during installation:
 Access to the path 'C:\ProgramData\chocolatey\lib\Ghostscript.app\tools\gs9550w64.exe' is denied.

Chocolatey installed 5/6 packages. 1 packages failed.
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).

Installed:
 - imagemagick.app v7.1.0.10
 - dejavufonts v2.37
 - imagemagick v7.1.0.10
 - vcredist2010 v10.0.40219.32503
 - ghostscript.app v9.55.0

Failures
 - ghostscript (exited 1) - ghostscript not installed. An error occurred during installation:
 Access to the path 'C:\ProgramData\chocolatey\lib\Ghostscript.app\tools\gs9550w64.exe' is denied.
Error: Process completed with exit code 1.
```

**Comment by PhilterPaper** [comment]
OK, I'll leave test.yml alone for now, so long as I don't have frequent commits to make. Just to repeat, tiff.t (and other t-tests) work fine on Windows (my local Strawberry installation); it's just on Github that it's failing for Windows.

**Comment by PhilterPaper** [comment]
I had to go ahead and disable the full-windows section of test.yml. Thanks for the effort, and if you can come up with a definite fix for it, that's great, but I don't think it's any great loss to not fully test on Windows. The code (.pm files) is checked on the Ubuntu test, so only certain parts of the Windows _build_ process are really being skipped, and I should be catching anything there when I build and run on my local Windows system.

**Comment by PhilterPaper** [comment]
The GD owner (see https://rt.cpan.org/Public/Bug/Display.html?id=141125) suggests using cpm rather than cpan to install prereqs. By any chance have you looked into this?

**Comment by carygravel** [comment]
No. And I only use cpan when testing with the Windows runners on Github, as for the Linux runners, the Perl dependencies are more quickly resolved using apt.

**Comment by PhilterPaper** [comment]
Attached is the GD test.yml as a possible model for PDF::Builder (I haven't tried anything). Just FYI.
[GD-test.yml.txt](https://github.com/PhilterPaper/Perl-PDF-Builder/files/8037872/GD-test.yml.txt) rename to test.yml.

**Comment by PhilterPaper** [comment]
I was just looking at something in test.yml (GitHub build/test on commit) and noticed that you require a ppa: **jeffreyratcliffe/ppa** for the full Ubuntu build. I'm a bit concerned that it is something that is not "stock" code, and thus PDF::Builder could end up passing on GitHub, but fail on someone's install. What is in this ppa? Something to do with gscan2pdf? I don't want tests depending on private repositories in the long term -- they should use public tools (private tools are OK over a short-term project). I tried commenting out the ppa line and full-ubuntu failed, so it's depending on _something_.

By the way, lint, full-, and basic-windows all started failing at one point or another, so I have disabled them for now. I don't think Windows testing is critical, as I build and run on Windows all the time, although it would be nice to have that working again. I run perlcritic against my code, so a lint isn't all that critical.

**Comment by carygravel** [comment]
`ppa: jeffreyratcliffe/ppa` is used so that it doesn't have to build Graphics::TIFF from source. I [uploaded it to Debian](https://tracker.debian.org/pkg/libgraphics-tiff-perl) as soon as I released the first version in September 2020. Ubuntu automatically uploaded it from Debian, so the first Ubuntu release with Graphics::TIFF was Impish (AKA 21.04), [released in April 2021](https://launchpad.net/ubuntu/+source/libgraphics-tiff-perl). The github runners are still using Ubuntu 20.04, as that is the current long term support release (LTS).

The next Ubuntu LTS release is 22.04, which I assume github will switch to soon afterwards. i.e. the ppa will probably no longer be necessary sometime in May or June.

**Comment by PhilterPaper** [comment]
Sounds good, but doesn't Ubuntu et al. have libtiff already built? Is this just the "glue" XS code you're referring to, or is libtiff just too old a version (at the moment)? Anyway, I'm looking forward to being able to get rid of private repository prereqs. Please let me know when I can update test.yml to get rid of the ppa.

Over on the Windows side, since libtiff doesn't ship with Windows _itself_, can anything similar be done so that it doesn't have to build TIFF each time? Is the ppa style of things Ubuntu only? Strawberry Perl 5.22 (the current minimum for PDF::Builder) ships with libtiff 4.0.3, and 5.24 ships with 4.0.6. Are those too back-level? I vaguely recall needing 4.0.10 but I might be mistaken. **Some time after May of 2022 I plan to up the minimum Perl to 5.24.** Is it reasonable to have users manually obtain the libtiff from a later version of Strawberry Perl and overwrite their back-level version? I wonder if supplying a libtiff library for the GitHub Windows build (copied from Strawberry) would be permitted (by license), and even included with the PDF::Builder package? As for ActiveState and other old Perl distributions, I don't have anything to recommend if they want to stay with that. I have no experience with supplying canned library (e.g., libtiff.a) files that any build process can easily use. Do you have any experience with this?

I haven't played around in the TIFF-related build material for a while, so I may be forgetting or misremembering something that's already been discussed.

**Comment by carygravel** [comment]
Graphics::TIFF is just the XS wrapper, not libtiff itself. The ppa is only Ubuntu - not just only Linux.

Graphics::TIFF needs a minimum of libtiff 4.0.3, so what Strawberry Perl ships is fine.

If you wanted to provide something newer, given that the full Windows build starts by building libtiff using alien, you could take that build and turn it into an msi. If libtiff is available, then it is reasonable to expect Windows users to build Graphics::TIFF against it. Indeed, if you wanted to provide Graphics::TIFF binaries, you'd have to ship a set of binaries for every combination of libtiff version and Perl version.

**Comment by PhilterPaper** [comment]
> Graphics::TIFF is just the XS wrapper, not libtiff itself. 

Yeah, I was commenting on the Windows part of the GitHub build trying to build libtiff itself, which seems to use up a huge amount of time and is failing anyway. Do you suppose it would be possible to include libtiff.a from some recent Strawberry Perl version in the PDF::Builder GitHub repository, and have test.yml use it for the Windows full build? If that works, perhaps libtiff.a could be shipped with the PDF::Builder distribution, although I don't know how to untangle the various distributions (e.g., Strawberry Perl would not need it, but ActiveState and others might need it, and rather than using Alien to try to build it...).

Making sure the license allows this is one thing; making it work technically (needed only on non-Strawberry Windows) is quite another. Have you seen any releases which successfully do this?

**Comment by PhilterPaper** [comment]
See also #199 concerning the need for a good debugger.

**Comment by PhilterPaper** [comment]
Some of the TIFF images that PDF::Builder fails with appear to be only with Adobe Acrobat Reader (the free download). Although it is so widely used that it may be considered the Gold Standard, AAR seems to have some limitations, especially in support of various compression formats, that other Readers have no problem with. In such cases, all I can do is document the sorts of things AAR is found to have trouble with, and suggest that users try a better Reader if they plan to use such formats.
